KAFKA-18757: Create full-function SimpleAssignor to match KIP-932 description

adixitconfluent opened the pull request:
### About
The current `SimpleAssignor` in AK assigned all subscribed topic partitions to all the share group members. This does not match the description given in [KIP-932](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=255070434#KIP932:QueuesforKafka-TheSimpleAssignor). Here are the rules as mentioned in the KIP by which the assignment should happen. We have changed the step 3 implementation here due to the reasons [described](https://github.com/apache/kafka/pull/18864#issuecomment-2659266502) -

1. The assignor hashes the member IDs of the members and maps the partitions assigned to the members based on the hash. This gives approximately even balance.
2. If any partitions were not assigned any members by (1) and do not have members already assigned in the current assignment, members are assigned round-robin until each partition has at least one member assigned to it.
3. We combine the current and new assignment. (Original rule - If any partitions were assigned members by (1) and also have members in the current assignment assigned by (2), the members assigned by (2) are removed.)

### Tests
The added code has been verified with unit tests and the already present integration tests.

adixitconfluent committed:
Added new SimpleAssignor

adixitconfluent committed:
Added unit tests

adixitconfluent committed:
Correcting step 3 logic for assignment computation

adixitconfluent committed:
Fixed failing tests

adixitconfluent committed:
Minor improvement

adixitconfluent committed:
Fixing test failure

adixitconfluent committed:
enhanced test testShareGroupHeartbeatWithMultipleMembers

adixitconfluent committed:
Enhanced the logic when combining new assignment and current assignment

adixitconfluent committed:
Changed HashSet to LinkedHashSet to preserve order while calculating hash for current assignment + unit test

adixitconfluent committed:
Revert "Changed HashSet to LinkedHashSet to preserve order while calculating hash for current assignment + unit test"

This reverts commit 86a4c6f566bae580c23b7d0fa2c124195a4a566d.

adixitconfluent commented:
Hi @AndrewJSchofield @apoorvmittal10 , the step 3 described above is a little tricky to implement (since we can only know the current assignment, not whether it was calculated by step 1 or step 2). I have implemented a way to filter current assignment as required in step 3 in function `filterCurrentAssignment`, but it is incorrect for a few cases. Maybe step 3 needs more consideration (or a future PR), but can we please meanwhile review the PR in its current state. I can also get rid of step 3 as of now and implement it in a new PR as well. Let me know your thoughts.

adixitconfluent committed:
Changed HashSet to LinkedHashSet to preserve order while calculating hash for current assignment

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -77,6 +85,17 @@ public void testAssignWithEmptyMembers() {
+        groupSpec = new GroupSpecImpl(
+            Collections.emptyMap(),
+            HETEROGENEOUS,
+            Collections.emptyMap()
```
Map.of() now, same above and elsewhere.

adixitconfluent committed:
Addressed copilot's comment

adixitconfluent committed:
Ammended step 3 of the simple assignor rules + unit tests

adixitconfluent commented:
I have amended the implementation of the **step 3** of the assignment such that we will combine new and current assignment without revoking the assignments that were assigned by step 1 in the new assignment and have members in current assignment by step 2. This has been done to avoid the complexity in both the implementation and the run time complexity because as of now we can only get the current assignment while calculating the new assignment. We do not have a way to know with which step, a particular assignment happened in the current assignment. I do have a way with which we can recreate the step wise assignment using the current assignment but that involves sorting and unnecessary computation. Hence, I am deferring with that approach.
IMO, **step 3** helps in reducing the burden of certain members of the share groups. This can be achieved with the help of limiting the max no. of partitions assignment for every member([KAFKA-18788](https://issues.apache.org/jira/browse/KAFKA-18788)). Hence, the potential problem of burdening the share consumers will be addressed in a future PR.
PS - We shouldn't have any problem in merging the PR to trunk with the amendment I suggested since right now, we anyways assign all the topic partitions to all the share group members which would be leading to burdening the share consumers anyways.
cc- @AndrewJSchofield @apoorvmittal10

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -111,12 +258,26 @@ private Map<Uuid, Set<Integer>> computeTargetPartitions(
+
+    record TargetPartition(Uuid topicId, int partition) {
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            TargetPartition that = (TargetPartition) o;
+            return topicId.equals(that.topicId) && partition == that.partition;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(topicId, partition);
+        }
```
Why do we need to define them, shouldn't record class automatically provides comparison based on data itself?

adixitconfluent replied:
used `org.apache.kafka.server.common.TopicIdPartition`, so we don't need TargetPartition now.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
         if (subscribeTopicIds.isEmpty())
             return new GroupAssignment(Collections.emptyMap());
```
Can we please move to Map.of now?

AndrewJSchofield replied:
Should be able to use `Map.of()` rather than `Collections.emptyMap()` across this class.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -111,12 +258,26 @@ private Map<Uuid, Set<Integer>> computeTargetPartitions(
         return targetPartitions;
     }
+
+    record TargetPartition(Uuid topicId, int partition) {
```
Why we didn't use `org.apache.kafka.server.common.TopicIdPartition`, seems same?

adixitconfluent replied:
you're right, we can use it. I was unable to find a class that did not contain topic name. Thanks for pointing it out.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+    }
 
-            members.put(memberId, new MemberAssignmentImpl(targetPartitions));
+    // Get the current assignment for subscribed topic partitions to share group members.
```
Please write the method comments as javadoc.

AndrewJSchofield commented:
Marked failed test as flaky in https://github.com/apache/kafka/pull/18925.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -111,12 +258,26 @@ private Map<Uuid, Set<Integer>> computeTargetPartitions(
+    record TargetPartition(Uuid topicId, int partition) {
+
+        @Override
+        public boolean equals(Object o) {
```
I've surprised that you implemented `equals` and `hashCode`. I believe the default implementation provided by the use of `record` would be sufficient.

adixitconfluent replied:
used `org.apache.kafka.server.common.TopicIdPartition` instead of `TargetPartition`, so we don't need it.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -26,12 +26,17 @@
-import java.util.function.Function;
 import java.util.stream.Collectors;
 
 import static org.apache.kafka.coordinator.group.api.assignor.SubscriptionType.HOMOGENEOUS;
```
The class javadoc comment for the simple assignor is now wrong.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+    void memberHashAssignment(
+        List<TargetPartition> targetPartitions,
+        Collection<String> memberIds,
+        Map<TargetPartition, List<String>> assignment) {
```
nit: As a general formatting point, indenting the arguments by 4 spaces, and then leaving the closing `) {` on the same line as the final argument means that the arguments and the following code statements are indented identically with no gap between them. Please reformat so the statements and the arguments are distinct. Personally, I would either place `) {` on a separate line, or indent all of the arguments at the end of the method name.

apoorvmittal10 replied:
Can we please javadoc comments explaining why to pass the `Map<TargetPartition, List<String>> assignment` rather this method returning `Map<TargetPartition, List<String>> assignment`.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+
+    // Visible for testing.
+    void roundRobinAssignment(
+        Collection<String> members,
```
nit: Previously, you use `memberIds` and `memberId`, as opposed to `members` and `member` in this method. I think the former is clearer because `member` sounds like a reference to a member object, as opposed to just its id.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
 
-    private Map<Uuid, Set<Integer>> computeTargetPartitions(
+    // Visible for testing.
+    void memberHashAssignment(
```
This is one class for which comprehensive comments would be helpful for future maintainers.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. Only a partial review so far, but I've left some initial comments.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+        memberHashAssignment(targetPartitions, groupSpec.memberIds(), newAssignment);
+
+        // Step 2: Round-robin assignment for unassigned partitions which do not have members already assigned in the current assignment.
+        Set<TargetPartition> assignedPartitions = new HashSet<>(newAssignment.keySet());
```
Why do we need to create new HashSet?

adixitconfluent replied:
you're right, we don't need it

apoorvmittal10 reviewed (changes requested):
Thanks for the PR, took an initial look. Some comments.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+        Collection<String> members,
+        List<TargetPartition> partitions,
+        Map<TargetPartition, List<String>> assignment) {
+        Iterator<String> memberIterator = members.iterator();
+        for (TargetPartition targetPartition : partitions) {
+            if (!memberIterator.hasNext()) {
+                memberIterator = members.iterator();
+            }
+            String member = memberIterator.next();
```
What we are trying to do here, can you please write comments.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
         if (subscribeTopicIds.isEmpty())
```
This should be better named `subscribedTopicIds`.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+        Map<TargetPartition, List<String>> currentAssignment) {
+
+        Map<TargetPartition, List<String>> newAssignment = new HashMap<>();
+        // Step 1: Hash member IDs to partitions.
```
Blank line before this comment please.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+            partitions.forEach(partition -> topicToMemberSubscription.computeIfAbsent(partition.topicId(), k -> new LinkedHashSet<>()).add(member)));
+
+        Map<TargetPartition, List<String>> newAssignment = new HashMap<>();
+        // Step 1: Hash member IDs to partitions.
```
Blank line before this comment please.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+        Map<TargetPartition, List<String>> newAssignment = new HashMap<>();
+        // Step 1: Hash member IDs to partitions.
+        memberToPartitionsSubscription.forEach((member, partitions) ->
+            memberHashAssignment(partitions, Collections.singletonList(member), newAssignment));
```
`List.of`

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+        List<TargetPartition> unassignedPartitions = targetPartitions.stream()
+            .filter(targetPartition -> !assignedPartitions.contains(targetPartition))
+            .filter(targetPartition -> !currentAssignment.containsKey(targetPartition))
+            .collect(Collectors.toList());
```
The `.collect(Collectors.toList())` can be replaced with `.toList()`.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+
+        Map<String, Set<TargetPartition>> finalAssignment = new HashMap<>();
+
+        // When combining current assignment, we need to only consider the topics in current assignment that are also being
```
This code to combine the current assignment with the new assignment to create the final assignment can be in common I think.

adixitconfluent replied:
We can do it but there will be an unnecessary change of a data structure from `Set` to `Map` because we need a topics to member subscription for heterogeneous (`topicToMemberSubscription`), but for homogeneous a set of subscribed topics is enough (`subscribedTopicIds`). Hence, I'll either need to convert the set to map to create a common function. Since it is 4-5 lines of code only, hence I opted not to make it common.

adixitconfluent committed:
Addressed Andrew's and Apoorv's round 1 comments - part 1

adixitconfluent committed:
Addressed Andrew's and Apoorv's round 1 comments - part 2

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+        List<TopicIdPartition> targetPartitions,
+        Map<TopicIdPartition, List<String>> currentAssignment
+    ) {
+
```
nit: remove line break please

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
-    private Map<Uuid, Set<Integer>> computeTargetPartitions(
-        Set<Uuid> subscribeTopicIds,
+    /**
+     * This function updates assignment by hashing the member IDs of the members and maps the partitions assigned to the members based on the hash. This gives approximately even balance.
```
Please break it in 2 lines.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+    ) {
+        if (!targetPartitions.isEmpty())
+            for (String memberId : memberIds) {
+                int topicPartitionIndex = Math.abs(memberId.hashCode() % targetPartitions.size());
```
Query: We will not factor in any stickiness in assignments? I guess the future PRs will have a better version of stickiness, correct? Else there could be pre-fetched records in client and additional work to release records when assignment changes.
cc: @AndrewJSchofield

TaiJuWu replied:
Just question, should we use following hash function ?https://github.com/apache/kafka/blob/a0470726c44a26a1cbd0cb7f6c29b4e3e28a3af1/streams/src/main/java/org/apache/kafka/streams/state/internals/Murmur3.java#L37 

Maybe internal hash function will change over time / version?

adixitconfluent replied:
@apoorvmittal10 , IMO, stickiness will be affected when either there is a change in members/ change in subscribed topic partitions. One of the future PR will aim to `Improve simple assignor by having a max no. of partitions limit and min no. of partitions threshold assignment to members`. The description of [simple assignor rules ](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=255070434#KIP932:QueuesforKafka-TheSimpleAssignor)does not factor in stickiness, so I guess that should be addressed in a future PR as well.

@TaiJuWu, sorry I couldn't understand your comment. Do you mean that we shouldn't use `hashCode`, rather use something like `hash128` from `Murmur3`? Any reason for such suggestion?

TaiJuWu replied:
Sorry for unclear, my question is: 

Is it possible for different versions of Java to have different implementations of hashCode?
If so, using the built-in hashCode result in different assignment outcomes?

But I am not sure this is actually an issue.

adixitconfluent replied:
@TaiJuWu , hashCode can have different implementations in different versions of Java. But why do we want to ensure same outcomes across different JVMs. All i think we need is a consistent result in a single instance o JVM which is happening here. Wdyt?

TaiJuWu replied:
Make sense to me, thanks for your explanation.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+    ) {
+
+        Map<TopicIdPartition, List<String>> newAssignment = new HashMap<>();
+
```
nit: remove line break.

adixitconfluent replied:
I had added this line break as per review comment form Andrew https://github.com/apache/kafka/pull/18864#discussion_r1958257137

AndrewJSchofield replied:
Indeed. If we are going to have a blank line before each comment which is "Step X:", then there should be one here too.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+
+        Map<TopicIdPartition, List<String>> newAssignment = new HashMap<>();
+
+        // Step 1: Hash member IDs to partitions.
```
```suggestion
        // Step 1: Hash member IDs to topic partitions.
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+        memberHashAssignment(targetPartitions, groupSpec.memberIds(), newAssignment);
+
+        // Step 2: Round-robin assignment for unassigned partitions which do not have members already assigned in the current assignment.
+        Set<TopicIdPartition> assignedPartitions = newAssignment.keySet();
```
Why do we need separate variable to hold this? Can't we just call `newAssignment.containsKey` below?

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+     *                   function would be called multiple times for heterogeneous assignment.
+     */
+    void memberHashAssignment(
+        List<TopicIdPartition> targetPartitions,
```
```suggestion
        List<TopicIdPartition> unassignedPartitions,
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+     */
+    void roundRobinAssignment(
+        Collection<String> memberIds,
+        List<TopicIdPartition> targetPartitions,
```
targetPartitions seems confusing here.
```suggestion
        List<TopicIdPartition> unassignedPartitions/topicIdPartitions,
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+    /**
+     * This functions assigns topic partitions to members by round-robin approach and updates the existing assignment.
+     * @param memberIds - the member ids to which the topic partitions need to be assigned.
+     * @param targetPartitions - the subscribed topic partitions.
```
This seems incorrect, it should be the partitions which needs assignment or something.

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+
+    /**
+     * This functions assigns topic partitions to members by round-robin approach and updates the existing assignment.
+     * @param memberIds - the member ids to which the topic partitions need to be assigned.
```
```suggestion
     * @param memberIds - the member ids to which the topic partitions need to be assigned, should be non-empty
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+            if (!memberIdIterator.hasNext()) {
+                memberIdIterator = memberIds.iterator();
+            }
+            String member = memberIdIterator.next();
```
```suggestion
            String memberId = memberIdIterator.next();
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+        unassignedPartitions.keySet().forEach(unassignedTopic ->
+            roundRobinAssignment(topicToMemberSubscription.get(unassignedTopic), unassignedPartitions.get(unassignedTopic), newAssignment));
+
+        // Step 3: We combine current assignment and new assignment.
+        Map<String, Set<TopicIdPartition>> finalAssignment = new HashMap<>();
+
+        // When combining current assignment, we need to only consider the member topic subscription in current assignment
+        // which is being subscribed in the new assignment as well.
+        currentAssignment.forEach((targetPartition, members) -> members.forEach(member -> {
+            if (topicToMemberSubscription.getOrDefault(targetPartition.topicId(), Collections.emptySet()).contains(member))
+                finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
+        }));
+        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
+            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
```
Is this as per the KIP or a changed approach to have more sharing? If changed then can you please write details in comment for now.

adixitconfluent replied:
This is a changed approach. Added the comment `As per the KIP, we should revoke the assignments from current assignment for partitions that were assigned by step 1 in the new assignment and have members in current assignment by step 2. But we haven't implemented it to avoid the complexity in both the implementation and the run time complexity. This step was mentioned in the KIP to reduce the burden of certain members of the share groups. This can be achieved with the help of limiting the max no. of partitions assignment for every member(KAFKA-18788). Hence, the potential problem of burdening the share consumers will be addressed in a future PR.`

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+
+        // When combining current assignment, we need to only consider the member topic subscription in current assignment
+        // which is being subscribed in the new assignment as well.
+        currentAssignment.forEach((targetPartition, members) -> members.forEach(member -> {
```
```suggestion
        currentAssignment.forEach((topicIdPartition, members) -> members.forEach(member -> {
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+        Map<String, Set<TopicIdPartition>> finalAssignment = new HashMap<>();
+
+        // When combining current assignment, we need to only consider the topics in current assignment that are also being
+        // subscribed in the new assignment as well.
+        currentAssignment.forEach((targetPartition, members) -> {
+            if (subscribedTopicIds.contains(targetPartition.topicId()))
+                members.forEach(member -> {
+                    if (groupSpec.memberIds().contains(member))
+                        finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
+                });
+        });
+        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
+            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
```
Is it aligned with the approach defiend in step 3 of the KIP?

adixitconfluent replied:
This is a changed approach. Added the comment `As per the KIP, we should revoke the assignments from current assignment for partitions that were assigned by step 1 in the new assignment and have members in current assignment by step 2. But we haven't implemented it to avoid the complexity in both the implementation and the run time complexity. This step was mentioned in the KIP to reduce the burden of certain members of the share groups. This can be achieved with the help of limiting the max no. of partitions assignment for every member(KAFKA-18788). Hence, the potential problem of burdening the share consumers will be addressed in a future PR.`

apoorvmittal10 replied:
The comment here talks about what's avoided but not about what's implemented. Can you please add that as well.

adixitconfluent replied:
It is mentioned [here](https://github.com/apache/kafka/pull/18864/files#diff-57e6fa203281f37b0be19bfce8db7583124e155e93e60b9744396f74db5932daR144) that` Step 3: We combine current assignment and new assignment.` Hence I don't think any more comment is required. If you want me to add anything else, please let me know.

apoorvmittal10 replied:
Ok, if you think it's sufficient then it is. In case you want to update PR description for step 3 then it ll be good.

adixitconfluent replied:
changed the description of the PR.

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -320,12 +365,377 @@ public void testAssignWithOneMemberNoAssignedTopicHeterogeneous() {
+        String member4 = "member4";
+        // hashcode for "AaAaAaAa" is -540425984 to test with negative hashcode.
+        String member5 = "AaAaAaAa";
+        List<String> members = Arrays.asList(member1, member2, member3, member4, member5);
```
```suggestion
        List<String> members = List.of(member1, member2, member3, member4, member5);
```

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -320,12 +365,377 @@ public void testAssignWithOneMemberNoAssignedTopicHeterogeneous() {
+        TopicIdPartition partition1 = new TopicIdPartition(TOPIC_1_UUID, 0);
+        TopicIdPartition partition2 = new TopicIdPartition(TOPIC_2_UUID, 0);
+        TopicIdPartition partition3 = new TopicIdPartition(TOPIC_3_UUID, 0);
+        List<TopicIdPartition> partitions = Arrays.asList(partition1, partition2, partition3);
```
```suggestion
        List<TopicIdPartition> partitions = List.of(partition1, partition2, partition3);
```

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,210 @@ private GroupAssignment assignHomogenous(
+     * @param targetPartitions - the subscribed topic partitions.
+     * @param memberIds - the member ids to which the topic partitions need to be assigned.
+     * @param assignment - the existing assignment by topic partition. We need to pass it as a parameter because this
+     *                   function would be called multiple times for heterogeneous assignment.
+     */
+    void memberHashAssignment(
```
```suggestion
     *                   function would be called multiple times for heterogeneous assignment.
     */
     // Visible for testing
    void memberHashAssignment(
```

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -320,12 +365,377 @@ public void testAssignWithOneMemberNoAssignedTopicHeterogeneous() {
+
+        Map<TopicIdPartition, List<String>> expectedAssignment = new HashMap<>();
+        expectedAssignment.put(partition1, List.of(member3));
+        expectedAssignment.put(partition2, Arrays.asList(member1, member4));
```
Why not List.of as in line above?

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -320,12 +365,377 @@ public void testAssignWithOneMemberNoAssignedTopicHeterogeneous() {
+        assignment.put(partition1, List.of(member1));
+
+        assignor.roundRobinAssignment(members, unassignedPartitions, assignment);
+        Map<TopicIdPartition, List<String>> expectedAssignment = new HashMap<>();
+        expectedAssignment.put(partition1, List.of(member1));
+        expectedAssignment.put(partition2, List.of(member1));
+        expectedAssignment.put(partition3, List.of(member2));
+        expectedAssignment.put(partition4, List.of(member1));
```
nit: can use Map.of(.....)

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -320,12 +365,377 @@ public void testAssignWithOneMemberNoAssignedTopicHeterogeneous() {
+            2
+        ));
+
+        Map<String, MemberSubscriptionAndAssignmentImpl> members1 = new TreeMap<>();
```
Question: Why treemap and cosidering order? The hash and assignment should be predictable in tests?

adixitconfluent replied:
actually we don't need to consider order here. I have just used a `HashMap` here and at other places in the code.

apoorvmittal10 reviewed (commented):
Some comments, though seems PR as good starting point and we migh improve on better partition stickiness while revoking and assigning partitions.

adixitconfluent committed:
Addressed Apoorv's round 2 comments

adixitconfluent committed:
Addressed Apoorv's round 2 comments - part 2

apoorvmittal10 commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,224 @@ private GroupAssignment assignHomogenous(
+        return assignment;
+    }
+
+    private GroupAssignment newAssignmentHomogeneous(
```
Would be good to have javadoc comments for these critical methods.

adixitconfluent replied:
since it is a private method and [javadocs are generally written for public methods](https://stackoverflow.com/questions/21632086/javadoc-for-private-protected-methods#:~:text=Nope%2C%20you%20shouldn't%20write,comments%20for%20non%2Dobvious%20logic.), I have avoided writing them for private method here.

apoorvmittal10 replied:
Yeah, but the code does a lot and it will be referenced while doing a refactor. Good to write them for self and others touching the code.

apoorvmittal10 commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -336,4 +745,31 @@ private void assertAssignment(
         }
     }
+
+    private void assertAssignment(
+        Map<TopicIdPartition, List<String>> expectedAssignment,
+        Map<TopicIdPartition, List<String>> computedAssignment
+    ) {
+        assertEquals(expectedAssignment.size(), computedAssignment.size());
+        expectedAssignment.forEach((topicIdPartition, members) -> {
+            List<String> computedMembers = computedAssignment.getOrDefault(topicIdPartition, List.of());
+            assertEquals(members.size(), computedMembers.size());
+            members.forEach(member -> assertTrue(computedMembers.contains(member)));
+        });
+    }
```
Question: the method is required because the list order might not be same?

adixitconfluent replied:
yes

apoorvmittal10 reviewed (commented):
Mostly looks good to me, can you please share the numbers with 16 partitions and 25 share consumers, with and without the PR.

adixitconfluent commented:
> can you please share the numbers with 16 partitions and 25 share consumers, with and without the PR.

@apoorvmittal10 , here are the numbers for 16 partitions and 25 share consumers - 

With PR - 
1 million records of size 1024 each - 17.5 seconds
5 million records of size 1024 each -  91 seconds

Without PR - 
1 million records of size 1024 each - 14.1 seconds
5 million records of size 1024 each - 72.2 seconds

As mentioned above as well, this PR reduces the sharing of topic partitions from the assignor, hence the decline in performance is expected. With the future PRs, the performance should reach an optimum number.

adixitconfluent committed:
Addressed Apoorv's round 3 comments

apoorvmittal10 commented:
> > can you please share the numbers with 16 partitions and 25 share consumers, with and without the PR.
> 
> @apoorvmittal10 , here are the numbers for 16 partitions and 25 share consumers -
> 
> With PR - 1 million records of size 1024 each - 17.5 seconds 5 million records of size 1024 each - 91 seconds
> 
> Without PR - 1 million records of size 1024 each - 14.1 seconds 5 million records of size 1024 each - 72.2 seconds
> 
> As mentioned above as well, this PR reduces the sharing of topic partitions from the assignor, hence the decline in performance is expected. With the future PRs, the performance should reach an optimum number.

Just to clarify, how was the partition allocation with the current PR code. Also if members are removed and added then there would be more sharing of partitions as per the combine logic in the PR, correct? Will it affect the performance?

adixitconfluent commented:
> > > can you please share the numbers with 16 partitions and 25 share consumers, with and without the PR.
> > 
> > 
> > @apoorvmittal10 , here are the numbers for 16 partitions and 25 share consumers -
> > With PR - 1 million records of size 1024 each - 17.5 seconds 5 million records of size 1024 each - 91 seconds
> > Without PR - 1 million records of size 1024 each - 14.1 seconds 5 million records of size 1024 each - 72.2 seconds
> > As mentioned above as well, this PR reduces the sharing of topic partitions from the assignor, hence the decline in performance is expected. With the future PRs, the performance should reach an optimum number.
> 
> Just to clarify, how was the partition allocation with the current PR code. Also if members are removed and added then there would be more sharing of partitions as per the combine logic in the PR, correct? Will it affect the performance?

right now, most of the members had 1-2 topic partitions allocated to them excepted 1-2 members which had a good 12-14 partitions assigned to them.
Yes, if members are removed and added then there would be more sharing of partitions as per the combine logic in the PR. Given the small size of 1-5 million records, it should improve the performance.

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/ShareGroupHeartbeatRequestTest.scala:
```diff
@@ -261,13 +255,23 @@ class ShareGroupHeartbeatRequestTest(cluster: ClusterInstance) {
       // Verify the response.
       assertEquals(3, shareGroupHeartbeatResponse.data.memberEpoch)
 
+      var partitionsAssigned: Set[Integer] = Set()
+      for (topicPartition <- topicPartitionsAssignedToMember1.asScala) {
+        partitionsAssigned = partitionsAssigned ++ topicPartition.partitions().asScala
+      }
+      for (topicPartition <- topicPartitionsAssignedToMember2.asScala) {
+        partitionsAssigned = partitionsAssigned ++ topicPartition.partitions().asScala
+      }
+      // Verify all the 3 topic partitions for "foo" have been assigned to at least 1 member.
+      assertEquals(Set(0, 1, 2), partitionsAssigned)
```
Instead of converting to scala the code can be written as:

```
      val partitionsAssigned: util.Set[Integer] = new util.HashSet[Integer]()
      topicPartitionsAssignedToMember1.forEach(topicPartition => {
        partitionsAssigned.addAll(topicPartition.partitions())
      })
      topicPartitionsAssignedToMember2.forEach(topicPartition => {
        partitionsAssigned.addAll(topicPartition.partitions())
      })
      // Verify all the 3 topic partitions for "foo" have been assigned to at least 1 member.
      assertEquals(util.Set.of(0, 1, 2), partitionsAssigned)

apoorvmittal10 reviewed (approved):
LGTM, given the code of the simple assignor will change in future PRs. One comment to address.

adixitconfluent committed:
Addressed Apoorv's comment around test

TaiJuWu commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -224,11 +253,15 @@ public void testAssignWithThreeMembersThreeTopicsHeterogeneous() {
         members.put(MEMBER_A, new MemberSubscriptionAndAssignmentImpl(
             Optional.empty(),
             Optional.empty(),
-            Set.of(TOPIC_1_UUID, TOPIC_2_UUID),
```
nit: is this change needed?

adixitconfluent replied:
I deliberately initialized it as a `LinkedHashSet` because I wanted ordering of topics, so that I can write an assert for the expected assignment. Hence, I had to make this change.

TaiJuWu replied:
Thanks for your explanation.

TaiJuWu commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignorTest.java:
```diff
@@ -239,18 +272,20 @@ public void testAssignWithThreeMembersThreeTopicsHeterogeneous() {
             Optional.empty(),
             Optional.empty(),
-            Set.of(TOPIC_2_UUID, TOPIC_3_UUID),
+            memberCTopicsSubscription,
```
nit: is this change needed?

adixitconfluent replied:
I deliberately initialized it as a `LinkedHashSet` because I wanted ordering of topics, so that I can write an assert for the expected assignment. Hence, I had to make this change.

TaiJuWu reviewed (approved):
LGTM. Just two nit for question but they are not very important.

AndrewJSchofield commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java:
```diff
@@ -67,42 +71,239 @@ private GroupAssignment assignHomogenous(
+        // the burden of certain members of the share groups. This can be achieved with the help of limiting the max
+        // no. of partitions assignment for every member(KAFKA-18788). Hence, the potential problem of burdening
+        // the share consumers will be addressed in a future PR.
+
```
Doesn't the following do the job a bit better?
```
        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
                finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
        currentAssignment.forEach((targetPartition, members) -> {
            if (subscribedTopicIds.contains(targetPartition.topicId())) {}
                members.forEach(member -> {
                    if (groupSpec.memberIds().contains(member) && !newAssignment.containsKey(targetPartition))
                        finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
                });
        });
```

The problem with the code as it currently exists is that it assigns all partitions to the first member, and then as other members join, it leaves all partitions with the first member in spite of assigning the partitions to the other members.

What the snippet above does is essentially give precedence to the new assignment, and only copies over information from the current assignment which augments the new assignment. It's still not perfect because the round-robin nature of the reassignment is not sophisticated enough, but I think it's probably better.

adixitconfluent replied:
makes sense. This will help reduce burdening of members, though it affects the stickiness of assignments now since we are revoking the assignments from current assignment. Now, we'll need to think of a way for optimum sharing in the future PRs. I have made this change.

adixitconfluent committed:
Addressed Andrew's comments

AndrewJSchofield reviewed (approved):
lgtm. Needs a bit more refinement, but this is a good start.
